fem/estimators.hpp

```diff
@@ -396,6 +398,148 @@ class LpErrorEstimator : public ErrorEstimator
    virtual ~LpErrorEstimator() {}
 };
 
+#ifdef MFEM_USE_MPI
+/** @brief The KellyErrorEstimator class provides a fast error indication
+    strategy for smooth scalar parallel problems.
+
+    The Kelly error indicator is based on the following papers:
+
+    Kelly, D. W., et al. "A posteriori error analysis and adaptive processes in
+    the finite element method: Part I—Error analysis." International journal for
+    numerical methods in engineering 19.11 (1983): 1593-1619.
+
+    De SR Gago, J. P., et al. "A posteriori error analysis and adaptive
+    processes in the finite element method: Part II—Adaptive mesh refinement."
+    International journal for numerical methods in engineering 19.11 (1983):
+    1621-1656.
+
+    It can be roughly described by:
+        ||∇(u-uₕ)||ₑ ≦ √( C hₑ ∑ₖ (hₖ ∫ |J[∇uₕ]|²) dS )
+    where "e" denotes an element, ||⋅||ₑ the corresponding local norm and k the
+    corresponding faces. u is the analytic solution and uₕ the discretized solution.
+    hₖ and hₑ are factors dependend on the face and element geometry. J is the jump
+    function, i.e. the difference between the limits at each point for each side of
+    the face. A custom method to compute hₖ can be provided. It is also possible to
+    estimate the error only on a subspace by feeding this class an attribute array
+    describing the subspace.
+
+    @note This algorithm is only for Poisson problems a proper error esimator.
+    The current implementation does not reflect this, because the "C" factor is not
+    included.
+    It further assumes that the approximation error at the boundary is small enough,
+    as the implementation ignores boundary faces.
+*/
+class KellyErrorEstimator final : public ErrorEstimator
+{
+private:
+   int current_sequence = -1;
+
+   Vector error_estimates;
+
+   double total_error = 0.0;
+
+   Array<int> attributes;
+
+   /** @brief A method to compute hₖ on per-element basis.
+
+       This method weights the error approximation on the element level.
+
+       Defaults to hₖ=1.0.
+   */
+   std::function<double(ParMesh*, const int)> compute_element_coefficient;
```

Comment:
Maybe it would be good to typedef these callbacks (something like `ElementCoefFn` or so)